libsql-server/src/schema/migration.rs

```diff
@@ -0,0 +1,121 @@
+use rusqlite::Savepoint;
+
+use crate::connection::program::{Program, Vm};
+use crate::namespace::meta_store::MigrationTask;
+use crate::query_result_builder::QueryResultBuilder;
+
+use super::{Error, MigrationTaskStatus};
+
+pub fn setup_migration_table(conn: &mut rusqlite::Connection) -> Result<(), Error> {
+    conn.execute(
+        "CREATE TABLE IF NOT EXISTS __libsql_migration_tasks (
+            job_id INTEGER PRIMARY KEY,
+            status INTEGER,
+            migration TEXT NOT NULL,
+            error TEXT
+        )",
+        (),
+    )?;
+
+    Ok(())
+}
+
+pub fn enqueue_migration_task(
+    conn: &rusqlite::Connection,
+    task: &MigrationTask,
+    migration: &Program,
+) -> Result<(), Error> {
+    let migration = serde_json::to_string(migration).unwrap();
+    conn.execute(
+        "INSERT INTO __libsql_migration_tasks (job_id, status, migration) VALUES (?, ?, ?)",
+        (task.job_id(), *task.status() as u64, &migration),
+    )?;
+    Ok(())
+}
+
+pub fn step_migration_task_run(
+    conn: &rusqlite::Connection,
+    task: &MigrationTask,
+) -> Result<(), Error> {
+    conn.execute(
+        "
+        UPDATE __libsql_migration_tasks
+        SET status = ?
+        WHERE job_id = ? AND status = ?
+        ",
+        (
+            task.job_id(),
+            MigrationTaskStatus::Run as u64,
+            MigrationTaskStatus::DryRunSuccess as u64,
+        ),
+    )?;
+
+    Ok(())
+}
+
+pub fn perform_migration<B: QueryResultBuilder>(
+    conn: &mut rusqlite::Transaction,
+    migration: &Program,
+    dry_run: bool,
+    builder: B,
+) -> (Result<B, Error>, MigrationTaskStatus) {
+    // todo error handling is sketchy, improve
+    let mut savepoint = conn.savepoint().unwrap();
```

Comment:
I guess this is also a place for bottomless savepoint (see #1085) in order to guarantee, that in case of crashes we can restore db to a state where the dry run succeeded. However in a current API it operates on the `Database` type not a `Connection`.